portal-wire-protocol.md

```diff
@@ -0,0 +1,163 @@
+# Portal Wire Protocol
+
+The Portal wire protocol is the base p2p protocol by which nodes communicate.
+
+The different Portal networks **MAY** use this base protocol, but they **MUST** remain separated per network.
+This is done at the [Node Discovery Protocol v5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#talkreq-request-0x05) layer, by providing a different protocol bytestring, per network, in the `TALKREQ` message.
+
+The protocol bytestring in the `TALKREQ` message **MUST** have "portal-" as prefix then followed by the network name (e.g. "portal-state").
+
+Each network can specify which specific messages from the protocol it supports.
+
+## Messages
+
+All messages in the protocol are transmitted using the `TALKREQ` and `TALKRESP` messages from the base protocol.
+
+All messages have a `message_id` and `encoded_message` that are concatenated to form the `payload` for either a `TALKREQ` or `TALKRESP` message.
+
+```
+payload         := message_id | encoded_message
+message_id      := uint8
+encoded_message := bytes
+```
+
+The `encoded_message` component is the SSZ encoded payload for the message as is indicated by its `message_id`. Each message has its own `sedes` which dictates how it should be encoded and decoded.
+
+The SSZ sedes `ByteList` is used to alias `List[uint8, max_length=2048]`.
+
+All messages have a `type` which is either `request` or `response`.
+
+* `request` messages **MUST** be sent using a `TALKREQ`
+* `response` messages **MUST** be sent using a `TALKRESP`
+
+
+### Ping (0x01)
+
+Request message to check if a node is reachable, communicate basic information about our node, and request basic information about the recipient node.
+
+```
+message_id := 0x01
+type       := request
+sedes      := Container(enr_seq: uint64, data_radius: uint256, custom_payload: ByteList)
+```
+
+* `enr_seq`: The node's current sequence number of their ENR record.
+* `data_radius`: The nodes current maximum radius for data stored by this node.
+* `custom_payload`: Custom payload dependant on the network.
+
+
+### Pong (0x02)
+
+Response message to Ping(0x01)
+
+```
+message_id := 0x02
+type       := response
+sedes      := Container(enr_seq: uint64, data_radius: uint256)
+```
+
+* `enr_seq`: The node's current sequence number of their ENR record.
+* `data_radius`: The nodes current maximum radius for data stored by this node.
+* `custom_payload`: Custom payload dependant on the network.
+
+### Find Nodes (0x03)
+
+Request nodes from the peer's routing table at the given logarithmic distances.  The distance of `0` indicates a request for the peer's own ENR record.
+
+```
+message_id := 0x03
+type       := request
+sedes      := Container(distances: List[uint16, max_length=256])
```

Comment:
I don't feel strongly about this but there's a slightly cleaner schema we could use here:

```
Container(distances: List[uint8, max_length=256], include_enr: bool]
```

This would be trading the current definition of `distance == 0` to mean "send me your ENR" to simply being a boolean flag.  Saves us 1 byte per distance value at the cost of an extra byte for the boolean value.  I'm guessing that on average we won't actually save much space at all.  I am -0 on this idea but I wanted to surface it.

Reply:
Yeah, this would save us several bytes.

Not sure about it being cleaner as it would have to make us adjust the distances.

Reply:
I don't see the bytes savings as that compelling given that it makes interpretation of the distances more complex.  Make your best judgement here.  Both seem fine.